package/MDAnalysis/core/topologyattrs.py

```diff
@@ -32,21 +32,123 @@
 from ..lib import transformations, mdamath
 from ..exceptions import NoDataError, SelectionError
 from .topologyobjects import TopologyGroup
 from . import selection
-from .groups import (GroupBase, Atom, Residue, Segment,
+from .groups import (ComponentBase, GroupBase,
+                     Atom, Residue, Segment,
                      AtomGroup, ResidueGroup, SegmentGroup)
 
-_LENGTH_VALUEERROR = ("Setting {group} with wrong sized array. "
-                      "Length {group}: {lengroup}, length values: {lenvalues}")
+
+def _check_length(func):
+    """Wrapper which checks the length of inputs to set_X
+
+    set_X(self, group, values)
+
+    if group in (Atom, Residue, Segment):
+        values must be single values, ie int, float or string
+    else:
+        values must be single value OR same length as group
+
+    """
+    _SINGLE_VALUE_ERROR = ("Setting {cls} {attrname} with wrong sized input. "
+                           "Must use single value, length of supplied values: {lenvalues}.")
+    # Eg "Setting Residue resid with wrong sized input. Must use single value, length of supplied
+    # values: 2."
+
+    _GROUP_VALUE_ERROR = ("Setting {group} {attrname} with wrong sized array. "
+                          "Length {group}: {lengroup}, length of supplied values: {lenvalues}.")
+    # Eg "Setting AtomGroup masses with wrong sized array. Length AtomGroup: 100, length of
+    # supplied values: 50."
+
+    def _attr_len(values):
+        # quasi len measurement
+        # strings, floats, ints are len 0, ie not iterable
+        # other iterables are just len'd
+        if iterable(values):
+            return len(values)
+        else:
+            return 0  # special case
+
+    @functools.wraps(func)
+    def wrapper(attr, group, values):
+        val_len = _attr_len(values)
+
+        if isinstance(group, ComponentBase):
+            if not val_len == 0:
+                raise ValueError(_SINGLE_VALUE_ERROR.format(
+                    cls=group.__class__.__name__, attrname=attr.singular,
+                    lenvalues=val_len))
+        else:
+            if not (val_len == 0 or val_len == len(group)):
+                raise ValueError(_GROUP_VALUE_ERROR.format(
+                    group=group.__class__.__name__, attrname=attr.attrname,
+                    lengroup=len(group), lenvalues=val_len))
+        # if everything went OK, continue with the function
+        return func(attr, group, values)
+
+    return wrapper
+
+def _wronglevel_error(attr, group):
+    """Generate an error for setting attr at wrong level
+
+    attr : TopologyAttr that was accessed
+    group : Offending Component/Group
+
+    Eg:
+    setting mass of residue, gets called with attr=Masses, group=residue
+
+    raises a NotImplementedError with:
+    'Cannot set masses from Residue.  Use 'Residue.atoms.masses'
+
+    Mainly used to ensure consistent and helpful error messages
+    """
+    err_msg = "Cannot set {attr} from {cls}. "
```

Comment:
I think I would prefer this message template to be down in the function, closer to where it is used. But it really is a matter of taste.